posthog/models/channel_type/sql.py

```diff
@@ -0,0 +1,59 @@
+import json
+import os
+
+from posthog.settings import CLICKHOUSE_CLUSTER
+
+CHANNEL_DEFINITION_TABLE_SQL = f"""
+CREATE TABLE IF NOT EXISTS channel_definition ON CLUSTER {CLICKHOUSE_CLUSTER} (
+    domain String NOT NULL,
+    kind String NOT NULL,
+    domain_type String NULL,
+    type_if_paid String NULL,
+    type_if_organic String NULL,
+) ENGINE = MergeTree()
+ORDER BY domain;
+"""
+
+DROP_CHANNEL_DEFINITION_TABLE_SQL = f"DROP TABLE IF EXISTS channel_definition ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
+
+TRUNCATE_CHANNEL_DEFINITION_TABLE_SQL = f"TRUNCATE TABLE IF EXISTS channel_definition ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
+
+with open(os.path.join(os.path.dirname(__file__), "channel_definitions.json"), "r") as f:
+    CHANNEL_DEFINITIONS = json.loads(f.read())
+
+
+def format_value(value):
+    if value is None:
+        return "NULL"
+    elif isinstance(value, str):
+        return f"'{value}'"
+    else:
+        raise ValueError(f"Unknown value type {type(value)}")
+
+
+CHANNEL_DEFINITIONS_DATA_SQL = f"""
+INSERT INTO channel_definition (domain, kind, domain_type, type_if_paid, type_if_organic) VALUES
+{
+''',
+'''.join(map(lambda x: f'({" ,".join(map(format_value, x))})', CHANNEL_DEFINITIONS))},
+;
+"""
+
+# Use COMPLEX_KEY_HASHED, as we have a composite key
+CHANNEL_DEFINITION_DICTIONARY_SQL = """
+CREATE DICTIONARY IF NOT EXISTS channel_definition_dict (
+    domain String,
+    kind String,
+    domain_type Nullable(String),
+    type_if_paid Nullable(String),
+    type_if_organic Nullable(String)
+)
+PRIMARY KEY domain, kind
+SOURCE(CLICKHOUSE(TABLE 'channel_definition'))
+LIFETIME(MIN 1 MAX 3600)
```

Comment:
How does the lifetime work? Curious why these values

Reply:
The dictionary lives in memory, based off the clickhouse table. This is the min and max interval between updating the memory version of the dictionary. You have to provide a range because if you might have many clickhouse nodes and it would like to avoid hammering the source with all the nodes at the same time.

This data is very rarely updated (the source google doc is updated approximately once a year), so I just picked a large number with maximum flexibility. Not too large, as we don't want to have to wait half a day for changes to propagate, but one hour seemed fine, and we can run a manual command to update it if we need to.

It might not be a bad idea to increase the min, hopefully clickhouse is sensible and doesn't try to update every second, but setting it to 3000 would be safer

Reply:
Done